cvmfs/receiver/catalog_merge_tool.cc

```diff
@@ -0,0 +1,234 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include "catalog_merge_tool.h"
+
+#include "catalog.h"
+#include "hash.h"
+#include "logging.h"
+#include "manifest.h"
+#include "options.h"
+#include "upload.h"
+#include "util/posix.h"
+
+namespace {
+
+PathString MakeRelative(const PathString& path) {
+  std::string rel_path;
+  std::string abs_path = path.ToString();
+  if (abs_path[0] == '/') {
+    rel_path = abs_path.substr(1);
+  } else {
+    rel_path = abs_path;
+  }
+  return PathString(rel_path);
+}
+
+}  // namespace
+
+namespace receiver {
+
+CatalogMergeTool::ChangeItem::ChangeItem(ChangeType type,
+                                         const PathString& path,
+                                         const catalog::DirectoryEntry& entry1)
+    : type_(type),
+      path_(path),
+      xattrs_(),
+      entry1_(new catalog::DirectoryEntry(entry1)),
+      entry2_(NULL) {}
+
+CatalogMergeTool::ChangeItem::ChangeItem(ChangeType type,
+                                         const PathString& path,
+                                         const catalog::DirectoryEntry& entry1,
+                                         const XattrList& xattrs)
+    : type_(type),
+      path_(path),
+      xattrs_(xattrs),
+      entry1_(new catalog::DirectoryEntry(entry1)),
+      entry2_(NULL) {}
+
+CatalogMergeTool::ChangeItem::ChangeItem(ChangeType type,
+                                         const PathString& path,
+                                         const catalog::DirectoryEntry& entry1,
+                                         const catalog::DirectoryEntry& entry2)
+    : type_(type),
+      path_(path),
+      xattrs_(),
+      entry1_(new catalog::DirectoryEntry(entry1)),
+      entry2_(new catalog::DirectoryEntry(entry2)) {}
+
+CatalogMergeTool::ChangeItem::ChangeItem(const ChangeItem& other)
+    : type_(other.type_),
+      path_(other.path_),
+      xattrs_(other.xattrs_),
+      entry1_(new catalog::DirectoryEntry(*(other.entry1_))),
+      entry2_(other.entry2_ ? new catalog::DirectoryEntry(*(other.entry2_))
+                            : NULL) {}
+
+CatalogMergeTool::ChangeItem::~ChangeItem() {
+  if (entry1_) {
+    delete entry1_;
+    entry1_ = NULL;
```

Comment:
What I mean is that the `delete` operator can deal with `nullptr` input, and I'd think that we wouldn't need the `entry1_ = NULL` either because the `ChangeItem` is going away.